app/models/message.rb

```diff
@@ -23,4 +26,22 @@ def attach(files_to_attach, uploaded_by:)
       end
     end
   end
+
+  def top_k_chunks_grouped_by_document(count: 5)
+    return [] if content.blank? || documents.empty? || count <= 0
+
+    chunks = documents.flat_map(&:chunks)
+    chunks_with_embeddings = Chunk.where(id: chunks.map(&:id))
+    top_chunks = chunks_with_embeddings.nearest_neighbors(:embedding, embedding, distance: :cosine).first(count)
```

Comment:
If you add:

`has_many :chunks, through: :documents`, then this can be done in 1 query and I think simplified code:

```
chunks.nearest_neighbors(:embedding, embedding, distance: :cosine).first(count).group_by(&:document)
```